source/compiler-core/slang-spirv-core-grammar.cpp

```diff
@@ -0,0 +1,330 @@
+#include "slang-spirv-core-grammar.h"
+
+#include "../core/slang-rtti-util.h"
+#include "../core/slang-string-util.h"
+#include "slang-json-native.h"
+#include "slang-core-diagnostics.h"
+#include <limits>
+
+namespace Slang
+{
+using SpvWord = uint32_t;
+
+//
+// Structs which mirror the structure of spirv.core.grammar.json
+//
+// Commented members are those which currently don't use
+struct InstructionPrintingClass
+{
+    UnownedStringSlice tag;
+    UnownedStringSlice heading;
+};
+SLANG_MAKE_STRUCT_RTTI_INFO(
+    InstructionPrintingClass,
+    SLANG_RTTI_FIELD(tag),
+    SLANG_OPTIONAL_RTTI_FIELD(heading)
+);
+
+struct Operand
+{
+    UnownedStringSlice kind;
+    UnownedStringSlice quantifier;
+    // UnownedStringSlice name;
+};
+SLANG_MAKE_STRUCT_RTTI_INFO(
+    Operand,
+    SLANG_RTTI_FIELD(kind),
+    SLANG_OPTIONAL_RTTI_FIELD(quantifier)
+    //SLANG_RTTI_FIELD(name),
+);
+
+struct Instruction
+{
+    UnownedStringSlice opname;
+    UnownedStringSlice class_;
+    SpvWord opcode;
+    List<UnownedStringSlice> capabilities;
+    List<Operand> operands;
+};
+SLANG_MAKE_STRUCT_RTTI_INFO(
+    Instruction,
+    SLANG_RTTI_FIELD(opname),
+    SLANG_RTTI_FIELD_IMPL(class_, "class", 0),
+    SLANG_RTTI_FIELD(opcode),
+    SLANG_OPTIONAL_RTTI_FIELD(capabilities),
+    SLANG_OPTIONAL_RTTI_FIELD(operands)
+);
+
+struct Enumerant
+{
+    UnownedStringSlice enumerant;
+    JSONValue value;
+    List<UnownedStringSlice> capabilities;
+    // List<Operand> parameters;
+    // UnownedStringSlice version;
+    // UnownedStringSlice lastVersion;
+    // List<UnownedStringSlice> extensions;
+};
+SLANG_MAKE_STRUCT_RTTI_INFO(
+    Enumerant,
+    SLANG_RTTI_FIELD(enumerant),
+    SLANG_RTTI_FIELD(value),
+    SLANG_OPTIONAL_RTTI_FIELD(capabilities),
+    // SLANG_OPTIONAL_RTTI_FIELD(parameters),
+    // SLANG_OPTIONAL_RTTI_FIELD(version),
+    // SLANG_OPTIONAL_RTTI_FIELD(lastVersion),
+    // SLANG_OPTIONAL_RTTI_FIELD(extensions)
+);
+
+struct OperandKind
+{
+    UnownedStringSlice category;
+    UnownedStringSlice kind;
+    List<Enumerant> enumerants;
+};
+SLANG_MAKE_STRUCT_RTTI_INFO(
+    OperandKind,
+    SLANG_RTTI_FIELD(category),
+    SLANG_RTTI_FIELD(kind),
+    SLANG_OPTIONAL_RTTI_FIELD(enumerants)
+);
+
+struct SPIRVSpec
+{
+    // List<UnownedStringSlice> copyright;
+    // UnownedStringSlice magic_number;
+    // UInt32 major_version;
+    // UInt32 minor_version;
+    // UInt32 revision;
+    List<InstructionPrintingClass> instruction_printing_class;
```

Comment:
any reason to use a different naming convention here?

Reply:
Yeah, it matches what's in the spirv json exactly, which allows for concise macros just below.

Reply:
Got it!